### PR TITLE
forgot and reset password done

### PR DESCRIPTION
**Pull request by anees-rehman-brainx**

### CODE REVIEW

File: src/components/AddTodo/AddTodo.jsx [hunk 1]
```diff
@@ -60,14 +61,19 @@ const AddTodo = () => {
...
     //use effect
     useEffect(()=>{
-       const user = localStorage.getItem('user');
+      let subscribed = true; 
+      const user = localStorage.getItem('user');
        if(!user){
         navigate("/user/login")
        }    
-       if(todoId !== 0){
+       if(todoId !== 0 && subscribed){
```
Comment (ali-faiz-brainx) [inline src/components/AddTodo/AddTodo.jsx]: the condition should be !todo instead of todo !===0

File: src/services/userService.js [hunk 2]
```diff
@@ -15,4 +15,30 @@ export const login = async (user) => {
...
+    return response.data;
+}
+
+//reset password by link provided in mail
+export const resetPassword = async(data) => {
+    console.log("service hits")
+    const {id, token, password, confirmPassword} = data;
+    console.log(id, token, password)
+    const response = await axios.put(
+            `${API_URL}/user/reset_password/${id}/${token}`,
```
Comment (ali-faiz-brainx) [inline src/services/userService.js]: pass token in headers instead of query param [Security reasons]

File: src/components/ResetPassword.jsx [hunk 3]
```diff
@@ -1,13 +1,59 @@
...
+    return () => {
+      dispatch(reset())
+      subscribed = false;
+    }
+  },[isError, isSuccess])
+
+  //handle submit button action
   function handleSubmit(e){
+    e.preventDefault()
+    const passwordRegex = /^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[#?!@$%^&*-]).{8,}$/
```
Comment (ali-faiz-brainx) [inline src/components/ResetPassword.jsx]: use password regex from constant files, So it makes reusable in other components also if needed.

**Review by ali-faiz-brainx — CHANGES_REQUESTED** [review]

File: src/components/AddTodo/AddTodo.jsx [hunk 4]
```diff
@@ -60,14 +61,19 @@ const AddTodo = () => {
...
         navigate("/user/login")
        }    
-       if(todoId !== 0){
+       if(todoId && subscribed){
          dispatch(getTodoByTodoId(todoId));
        }
-    },[todoId])
+
+       return () => {
+        subscribed = true;
```
Comment (ali-iqbal-brainx) [inline src/components/AddTodo/AddTodo.jsx:74]: component should be subscribed=false on unmount(need to revisit your logic)

File: src/components/ForgotPassword.jsx [hunk 5]
```diff
@@ -1,12 +1,45 @@
...
-
   const [email, setEmail] = useState('');
+  const navigate = useNavigate();
+  const dispatch = useDispatch();
+
+  //getting data from redux store
+  const {isError, isSuccess, message} = useSelector(state => state.user)
+
+  useEffect(() => {
+    let subscribed = true;
```
Comment (ali-iqbal-brainx) [inline src/components/ForgotPassword.jsx:16]: use state instead of local variables

File: src/components/ForgotPassword.jsx [hunk 6]
```diff
@@ -25,7 +58,7 @@ const ForgotPassword = () => {
         </div>
 
         <div className="d-grid">
-          <button type="submit" className="btn btn-primary w-50 mx-auto my-2 fw-bold">
+          <button type="submit" className="btn btn-primary w-50 mx-auto my-2 fw-bold" onClick={handleSubmit}>
```
Comment (ali-iqbal-brainx) [inline src/components/ForgotPassword.jsx:61]: span component on multiple lines if it contains more than one attribute

File: src/components/ResetPassword.jsx [hunk 7]
```diff
@@ -1,13 +1,58 @@
...
   const [password, setPassword] = useState('');
   const [confirmPassword, setConfirmPassword] = useState('');
+  const navigate = useNavigate();
+  const dispatch = useDispatch();
+
+  
+  const {isError, isSuccess, message} = useSelector(state => state.user)
+
+  useEffect(() => {
+    let subscribed = true;
```
Comment (ali-iqbal-brainx) [inline src/components/ResetPassword.jsx:19]: use state instead of local variables

File: src/components/ResetPassword.jsx [hunk 8]
```diff
@@ -37,7 +82,7 @@ const ResetPassword = () => {
         </div>
         
         <div className="d-grid">
-          <button type="submit" className="btn btn-primary w-50 mx-auto my-2 fw-bold">
+          <button type="submit" className="btn btn-primary w-50 mx-auto my-2 fw-bold" onClick={handleSubmit}>
```
Comment (ali-iqbal-brainx) [inline src/components/ResetPassword.jsx:85]: span component on multiple lines

File: src/components/TodoList.jsx/TodoList.jsx [hunk 9]
```diff
@@ -15,8 +15,17 @@ const TodoList = () => {
   const todos = Array.from(todo);
 
   useEffect(() => {
-    dispatch(getAllTodosOfUser());
-  },[todo,navigate])
+    let subscribed = true;
```
Comment (ali-iqbal-brainx) [inline src/components/TodoList.jsx/TodoList.jsx:18]: use state here

**Review by ali-iqbal-brainx — COMMENTED** [review]
-component should be subscribed=false on unmount(need to revisit your logic)
-use state instead of local variables
-span component on multiple lines if it contains more than one attribute

**Pull request merged**